### PR TITLE
feat: add scaffolding CLI tool

### LINKED ISSUE #1330: Scaffold CLI tool for creating new projects
> _Fuels CLI integration requires `fuels.config.ts` file generation_ — #1291

Create a scaffolding tool for generating ready-to-go Fullstack Fuel dApps.

It should support templates and provide a single-command frictionless entry point for newcomers.

```console
npx create fuels@latest my-fuel-dapp
npx create fuels@latest my-fuel-dapp -t <template-name>
```

Companion frameworks example:

```console
npx create fuels@latest my-fuel-dapp -t next
npx create fuels@latest my-fuel-dapp -t nuxt
npx create fuels@latest my-fuel-dapp -t sveltekit
npx create fuels@latest my-fuel-dapp -t astro
npx create fuels@latest my-fuel-dapp -t qwik
```

Dead-simple instant quickstart:

```console
npx create fuels@latest my-fuel-quickstart
```

End goal:

- Single-command, no-brainer, frictionless entry point for Fullstack dApp engineers
- Get users prototyping Fuel dApps in _seconds_
- Fast development cycle

### LINKED ISSUE #1522: The fuels-forc / fuels-core packages don't install the binaries when used via bun
If we use bun to install the SDK, it doesn't install the binaries served via `@fuel-ts/fuels-forc` and `@fuel-ts/fuels-core` which are supposed to be installed via the `install` script.

We are not sure if this is a bug on our end, or an issue with bun. This needs to be investigated.

### PR DESCRIPTION
**Pull request by Dhaiwat10**
## Summary
This PR adds a new scaffold CLI tool that lets devs get started with a new full-stack Fuel project quickly by running:
`pnpm create fuels`

There are lot of file changes but most of them are just Node version corrections in all the `package.json`s. The ones you have to review are the new package `create-fuels` and the template inside the folder `template/nextjs`.

### Demo of the resulting scaffolded project

https://github.com/FuelLabs/fuels-ts/assets/39617427/271ed49b-ffdf-4285-841e-152b08b8d86c

^ This is what the users will get once they have run `pnpm create fuels`

### Supports:
- [x] npm
- [x] yarn
- [x] pnpm
- [x] bun

Closes #1330 
Closes #1522

### CODE REVIEW

File: packages/create-fuels/scripts/rewriteTemplatePackageJson.ts
Comment (arboleya) [inline packages/create-fuels/scripts/rewriteTemplatePackageJson.ts:33]: Because it's just a script, I'd go with something very minimal.

```suggestion
import { versions } from '@fuel-ts/versions';
import { readFileSync, writeFileSync } from 'fs';
import { join } from 'path';

const filepath = join(__dirname, '../templates/nextjs/package.json');

let contents = readFileSync(filepath, 'utf-8');

contents = contents.replace(/xprebuild/g, 'prebuild');
contents = contents.replace(/"fuels": "workspace:\*"/, `"fuels": "${versions.FUELS}"`);
contents = contents.replace(/"build": "next build"/, `"build": "next build"`);

writeFileSync(filepath, contents);
```

File: packages/create-fuels/tsup.config.ts
Comment (arboleya) [inline packages/create-fuels/tsup.config.ts:3]: Add the following to `internal/tsup/src/index.ts` immediately _before_ `export const indexBinAndCliConfig`:

```ts
export const bindAndCli: Options = {
  ...tsupDefaults,
  entry: {
    bin: 'src/bin.ts',
    cli: 'src/cli.ts',
  },
};
```

Then delete the `create-fuels/src/index.ts` file and:

```suggestion
import { binAndCli } from '@internal/tsup';

export default binAndCli;
```

File: templates/nextjs/fuels.config.ts
Comment (arboleya) [inline templates/nextjs/fuels.config.ts:8]: Nit.

```suggestion
export default createConfig({
  contracts: ['sway-contracts'],
  output: './src/sway-api',
});
```

File: templates/nextjs/package.json
Comment (arboleya) [inline templates/nextjs/package.json:6]: Nit.

```suggestion
    "xprebuild": "fuels build --deploy",
```

File: templates/nextjs/package.json
Comment (arboleya) [inline templates/nextjs/package.json:8]: Nit.

```suggestion
    "build": "npm run xprebuild && next build",
```

**Review by arboleya — CHANGES_REQUESTED** [review]
Left a couple of small suggestions.

I believe we should remove everything from the public directory, no?

**Comment by arboleya** [comment]
@Dhaiwat10 Also, we must have docs for this.

**Comment by Dhaiwat10** [comment]
Absolutely @arboleya, I've addressed your suggestions and I will take care of the docs and tests once I land in Istanbul! 🤝

**Review by arboleya — APPROVED** [review]

**Review by arboleya — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-release.yaml
Comment (Dhaiwat10) [reply .github/workflows/pr-release.yaml:11]: I will update this once the PR is reviewed & approved

Comment (arboleya) [reply .github/workflows/pr-release.yaml:11]: Wouldn't this dismiss all approvals?

File: packages/create-fuels/scripts/rewriteTemplatePackageJson.ts
Comment (nedsalk) [inline packages/create-fuels/scripts/rewriteTemplatePackageJson.ts:11]: Does this work? Line 9 replaces `xprebuild` with `prebuild` globally, but this one still has `xprebuild` in it (`...xprebuild && next build`)

**Review by nedsalk — COMMENTED** [review]
Clean stuff! I added some comments, but nothing special. I'll approve when they're resolved.

File: apps/docs/src/guide/quickstart/index.md
Comment (arboleya) [reply apps/docs/src/guide/quickstart/index.md:6]: For all doc pages _(not package `README.md`s)_, we should use this [tabbed component](https://github.com/FuelLabs/fuels-ts/pull/1470/files#diff-0c239f50914a185674b953ab25f3576acc94d54a43a4b3079a836a33e685afcf).

This PR should have standardized this usage across our docs:
 - https://github.com/FuelLabs/fuels-ts/pull/1470

File: packages/create-fuels/scripts/rewriteTemplatePackageJson.ts
Comment (Dhaiwat10) [reply packages/create-fuels/scripts/rewriteTemplatePackageJson.ts:11]: It works, this is the output in the template:
image

Comment (arboleya) [reply packages/create-fuels/scripts/rewriteTemplatePackageJson.ts:11]: 🕵️

File: templates/nextjs/fuels.config.ts
Comment (arboleya) [inline templates/nextjs/fuels.config.ts:6]: Should we follow the doc's [directory structure](https://fuellabs.github.io/fuels-ts/guide/cli/#getting-started)?
 - `sway-programs`
 - `sway-programs-api`

If not, then the docs should change.

Either way, both must be in accordance.

@FuelLabs/sdk-ts @FuelLabs/frontend Thoughts or preferences?

File: templates/nextjs/README.md
Comment (arboleya) [inline templates/nextjs/README.md:28]:
```suggestion
- [Fuel TS SDK docs](https://fuellabs.github.io/fuels-ts/)
```

File: templates/nextjs/src/pages/index.tsx
Comment (arboleya) [inline templates/nextjs/src/pages/index.tsx:25]: Nit. Sorry.

```suggestion
      // 0x1 is the private key of one of the fauceted accounts on your local Fuel node
      const wallet = Wallet.fromPrivateKey("0x01", provider); 
```

File: apps/docs/src/guide/quickstart/index.md
Comment (Dhaiwat10) [reply apps/docs/src/guide/quickstart/index.md:16]: @nedsalk @arboleya Great catch. Sorry, wasn't aware of this. Fixed!

**Review by arboleya — COMMENTED** [review]
Did a first pass, still need to use the CL tho. 🫡

File: templates/nextjs/fuels.config.ts
Comment (Dhaiwat10) [reply templates/nextjs/fuels.config.ts:6]: I think the structure mentioned in the docs is the same, it's just the names. 

I should have touched upon this in the PR description - I intentionally chose to use contracts over programs in this quickstart because most devs using this quickstart will be familiar with the term _contracts_ more than _programs_ in the context of a blockchain. It's good if we only introduce them to our contracts first and then somehow slowly introduce our _programs_ - i.e. scripts and predicates.

File: apps/docs/src/guide/quickstart/index.md
Comment (Dhaiwat10) [reply apps/docs/src/guide/quickstart/index.md]: @arboleya I added the link but the docs build fails:
image


:/ Maybe let's revert the link for now and then after this PR see if there's a way around this for localhost links

File: apps/docs/src/guide/quickstart/index.md
Comment (Torres-ssf) [inline apps/docs/src/guide/quickstart/index.md:8]: Does it need the project name?

This is the output that I am receiving

Screenshot 2023-12-18 at 18 45 16

I had to run `pnpm create fuels myProject` to make this work

File: templates/nextjs/fuels.config.ts
Comment (arboleya) [reply templates/nextjs/fuels.config.ts:6]: Could the default template receive a flag with what should be included?
 - `--contract` (default)
 - `--predicate`
 - `--script`
 - `--all`

File: apps/docs/src/guide/quickstart/index.md
Comment (Torres-ssf) [inline apps/docs/src/guide/quickstart/index.md:32]: When I `cd` to the project dir and run `pnpm install`, I am running into an error.
Screenshot 2023-12-18 at 19 09 06

I wonder if I am missing something.

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
Still going throught the review process:

File: apps/docs/src/guide/quickstart/index.md
Comment (Dhaiwat10) [reply apps/docs/src/guide/quickstart/index.md:8]: Please run the command with the pr release tag like I mentioned here: https://fuellabs.slack.com/archives/C04LVA1UF7X/p1702913792062199

If you don't use the pr release tag, it uses the first, old cached version

File: apps/docs/src/guide/quickstart/index.md
Comment (Dhaiwat10) [reply apps/docs/src/guide/quickstart/index.md:32]: Same as the previous comment. Please use the pr release tag and it should work fine

File: templates/nextjs/fuels.config.ts
Comment (arboleya) [reply templates/nextjs/fuels.config.ts:6]: Can you please open an issue for this?

Comment (Dhaiwat10) [reply templates/nextjs/fuels.config.ts:6]: #1562

Comment (arboleya) [reply templates/nextjs/fuels.config.ts:6]: Thank you.

**Comment by Dhaiwat10** [comment]
Closing this once since a lot of unwanted commits got included because of changing the base branch twice. Opening a new one soon.

**Pull request closed without merging**
